proposals/p1927.md

```diff
@@ -0,0 +1,148 @@
+# Slice generator
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1927)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [Library API](#library-api)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+Slices of arrays is a popular way to extract subarray of array. They also may be
+used for generating arrays.
+
+## Background
+
+There is no requirements for background.
+
+## Proposal
+
+Introduce slice generators.
+
+## Details
+
+Slice syntax is `a:b:s`. Here, `a` is the beginning of slice, `b` is the end of
+slice and `s` is the step in slice. In array indexing, `:s` can be omitted, then
+it means that step is 1. If `a` is omitted, slice starts from the beginning of
+array. The same situation is for `b`, but slice ends at the end of arrays. `:`
+is the same as `::`. If `s` is negative and `a` or `b` or both are skipped,
+instead of `a`, upper bound is and, instead of `b`, lower bound is. Slice has
+closed range.
+
+Compilation or runtime errors are:
+
+-   if `s` is zero;
+-   if `a` is greater than `b` when `s` is positive;
+-   if `a` is less than `b` when `s` is negative.
+
+`:` has the lowest priority for making possible `a[1:N-1]` without any brackets.
+
+Slice generator may return `i32`, `i64`, `i128` and unsigned versions of them.
```

Comment:
Can user-defined types be customized to work with this syntax? If so, how does that customization work? See also #1885.